File: Java/2048.java

// https://open.kattis.com/problems/2048
import java.util.Scanner;

public class 2048 {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int[][] board = new int[4][4];
        for (int x = 0, i = 0; x < 4; x++)
            for (String s : sc.nextLine().split(" ")) {
                board[x][i++] = Integer.parseInt(s);
                if (i == 4)
                    i = 0;
            }
        int direction = sc.nextInt();
        if (direction == 0)
            board = left(board);
        else if (direction == 1)
            board = up(board);
        else if (direction == 2)
            board = right(board);
        else
            board = down(board);

        print(board);

        sc.close();
    }

    private static int[][] left(int[][] arr) {
        // go through every row
        for (int r = 0; r < 4; r++) {
            // create a boolean array to keep track of which numbers/indexes
            // have combined to prevent a number from combining more than once
            boolean[] change = new boolean[4];
            // go through every column besides the first
            for (int c = 1; c < 4; c++)
                if (arr[r][c] != 0) {
                    // starting from index 1, shift each number and their boolean status as left as they can go
                    for (int pos = c; pos > 0; pos--) {
                        if (arr[r][pos - 1] == 0) {
                            arr[r][pos - 1] = arr[r][pos];
                            arr[r][pos] = 0;
                            boolean temp = change[pos - 1];
                            change[pos - 1] = change[pos];
                            change[pos] = temp;
                        } else if (arr[r][pos - 1] == arr[r][pos] && (!change[pos - 1] && !change[pos])) {
                            // basically if 2 numbers are equal then add them together
                            // and set their index to true to indicate that they combined
                            arr[r][pos - 1] *= 2;
                            arr[r][pos] = 0;
                            change[pos - 1] = true;
                        }
                    }
                }
        }
        return arr;
    }

    private static int[][] right(int[][] arr) {
        return rotate(rotate(left(rotate(rotate(arr)))));
    }

    private static int[][] up(int[][] arr) {
        return rotate(rotate(rotate(left(rotate(arr)))));
    }

    private static int[][] down(int[][] arr) {
        return rotate(left(rotate(rotate(rotate(arr)))));
    }

    private static int[][] rotate(int[][] arr) {
        int[][] temp = new int[4][4];
        for (int c = 3, row = 0; c >= 0; c--, row++)
            for (int r = 0, col = 0; r < 4; r++, col++)
                temp[row][col] = arr[r][c];
        return temp;
    }

    private static void print(int[][] arr) {
        for (int[] a : arr) {
            for (int s : a)
                System.out.print(s + " ");
            System.out.println();
        }
    }
}

File: C#/2048.cs
using System;

public class 2048 {
    public static void Main (string[] args) {
        int[, ] board = new int[4, 4];
        for (int r = 0, c = 0; r < 4; r++)
            foreach (string s in Console.ReadLine ().Split (' ')) {
                board[r, c++] = int.Parse (s);
                if (c == 4)
                    c = 0;
            }

        string dir = Console.ReadLine ();

        if (dir.Equals ("0"))
            print (left (board));
        else if (dir.Equals ("1"))
            print (up (board));
        else if (dir.Equals ("2"))
            print (right (board));
        else
            print (down (board));
    }

    private static int[, ] left (int[, ] arr) {
        for (int r = 0; r < 4; r++) {
            bool[] swap = new bool[4];
            for (int c = 1; c < 4; c++) {
                if (arr[r, c] == 0)
                    continue;
                for (int C = c; C > 0; C--) {
                    if (arr[r, C - 1] == 0) {
                        arr[r, C - 1] = arr[r, C];
                        arr[r, C] = 0;
                        bool temp = swap[C];
                        swap[C] = swap[C - 1];
                        swap[C - 1] = temp;
                    } else if (arr[r, C - 1] == arr[r, C] && !(swap[C] || swap[C - 1])) {
                        arr[r, C - 1] <<= 1;
                        arr[r, C] = 0;
                        swap[C - 1] = true;
                    }
                }
            }
        }
        return arr;
    }

    private static int[, ] right (int[, ] arr) {
        return rotate (rotate (left (rotate (rotate (arr)))));
    }

    private static int[, ] up (int[, ] arr) {
        return rotate (rotate (rotate (left (rotate (arr)))));
    }

    private static int[, ] down (int[, ] arr) {
        return rotate (left (rotate (rotate (rotate (arr)))));
    }

    private static int[, ] rotate (int[, ] arr) {
        int[, ] temp = new int[4, 4];
        for (int c = 3, R = 0; c >= 0; c--, R++)
            for (int r = 0, C = 0; r < 4; r++, C++)
                temp[R, C] = arr[r, c];
        return temp;
    }

    private static void print (int[, ] arr) {
        int i = 0;
        foreach (int x in arr) {
            Console.Write (x + " ");
            i++;
            if (i == 4) {
                Console.WriteLine ();
                i = 0;
            }
        }
    }
}
